Workaround for RPL forwarding inconsistency leading to dropped packet

arurke opened the pull request:
This is a workaround for issue #2285, where received packets that are addressed to ourselves and are forwarded inconsistently (according to [RFC6550 Section 11.2.2.2](https://www.rfc-editor.org/rfc/rfc6550#section-11.2.2.2)), will be dropped. This happens as the packet goes through the RPL extension-header processing twice: In the first pass it will be marked with the Rank-Error bit, and in the second pass the bit is seen and the packet is dropped.

This PR
1. Adds a workaround to the issue, ensuring the packet is not dropped.
2. ~~Adds a simulation test-case that showcases the issue. The test fails without the workaround, and passes with it.~~. The test for the issue was added separately in #2548 . This PR simply reverts the logic in the test such that it passes only if the issue is resolved.

I am calling this a workaround because it only mitigates the dropping of the packet (the issue still causes reset of DIO timer and erroneous RPL statistics). I have taken this conservative low-risk approach because I am no RPL/uIP expert and I can't really judge what is the proper solution - e.g. if we should rather avoid the double processing of extension-headers in such cases (and if so how to safely implement). As such I will have no trouble if we don't merge this and rather leave it for others to work onward from.

For further details see test-description in #2548 and issue #2285.

arurke commented:
@tinstructor Does this match your understanding and does the workaround work for you?

pjonsson commented on tests/15-rpl-classic/code-rank-error/Makefile:
```diff
@@ -0,0 +1,8 @@
+
+MAKE_ROUTING = MAKE_ROUTING_RPL_CLASSIC
+
+CONTIKI = ../../..
```
The top Makefile defines the top directory, so I don't think `$CONTIKI` is required to be defined by test cases anymore.

pjonsson commented on tests/15-rpl-classic/js/12-rpl-rank-error-mobility.dat:
```diff
@@ -0,0 +1,7 @@
+3 0.0 90 0
+3 180.0 30 20
+3 3600.0 30 20
+
```
Is the blank line required?

arurke replied:
I have this as a habit to ensure that the last line has terminating symbols. The idea is that some systems/tools/etc. out there defines a line as chars ending with terminators, and may misbehave if not present. Can't say I have done extensive research on this, but quick googling seems to say it is reasonable - happy to be convinced otherwise.

pjonsson replied:
This particular file is read by Cooja which isn't really portable at all (Linux/OS X and maybe Windows/FreeBSD). Even if Cooja was portable, one would want a fast host to run simulations, and to run things at all the host still needs a few gigs of RAM and a JVM. I think this narrows down the host platforms to Linux-like/OS X/Windows.

I'm not sure what to search for based on your description, isn't EOF a terminator?

Without being sure what class of bugs you are trying to avoid with the empty line, my gut feeling for the general case is that pipelining basic unix tools (awk/sed/grep/perl) becomes harder with files that contain spurious blank lines since the blank lines need to be filtered out from the main loop to not end up with an empty variable for the line (`$_` in perl, I have no idea what it is called in other languages).

arurke replied:
Sorry @pjonsson, I see now that I have added an actual blank line (as you of course wrote in the first comment). So there is now two LFs at the end of the file. This was not my intention, I wanted to ensure the last line had an LF. I got confused because in my local editor, if the last line has a LF, a blank line is shown beneath it. When I looked at the file here in github.com I saw a blank line and thought everything was as I wanted, but did not realize github does not show a separate blank line when the last line has LF - it was actually a second LF there.

> I'm not sure what to search for based on your description, isn't EOF a terminator?

I don't believe it is no, leaning on SO: https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline

arurke replied:
I have removed the stray line

pjonsson commented on tests/15-rpl-classic/code-rank-error/simple-node.c:
```diff
@@ -0,0 +1,103 @@
+PROCESS_THREAD(udp_process, ev, data)
+{
+  static struct etimer periodic_timer;
+  uip_ipaddr_t dest_addr;
```
Can this be declared in the while-loop?

arurke commented:
Thanks for the review @pjonsson!

pjonsson commented on tests/15-rpl-classic/js/12-rpl-rank-error.js:
```diff
@@ -0,0 +1,26 @@
+
+  else if(msg.contains("Received all packets")) {
+    log.log(time + " " + "node-" + id + " "+ msg + "\n");
+    all_transmissions_received = true
```
Does this test always terminate with a timeout? Is there any reason to not let it terminate with testOK() here?

arurke replied:
This was intentional, but after review I can't really see the rationale - there is plenty of time after the nodes move to catch any issues. I have simplified the script.

arurke commented on tests/15-rpl-classic/js/12-rpl-rank-error.js:
```diff
@@ -0,0 +1,26 @@
+all_transmissions_received = false;
+
+function check_success() {
```
No indeed, some kind of artifact, removed 👍

arurke commented:
I have now rebased and removed the tests that were added via #2548.

As mentioned I believe there is more work to be done on the issue (ref. my explanation above of why calling this a workaround) - at the same time I believe this PR is conservative and is unlikely to do any harm.

joakimeriksson commented:
@arurke - we did pull in Nicolas fix for this. So I do assume that we can close this? #2862

arurke commented:
@joakimeriksson Indeed! Replaced by #2862 

@nvt In case you aren't aware, there is one more similar issue+workaround in https://github.com/contiki-ng/contiki-ng/pull/2545

nvt commented:
@arurke Yes, I have seen it, and plan to take a closer look soon.
